cmd/bootstrap/cmd/seal.go

```diff
@@ -38,3 +39,22 @@ func constructRootResultAndSeal(
 
 	return result, seal
 }
+
+// rootEpochTargetEndTime computes the target end time for the given epoch, using the given config.
+// CAUTION: the variables `flagEpochTimingRefCounter`, `flagEpochTimingDuration`, and
+// `flagEpochTimingRefTimestamp` must contain proper values. You can either specify a value for
+// each config parameter or use the function `validateOrPopulateEpochTimingConfig()` to populate the variables
+// from defaults.
+func rootEpochTargetEndTime() uint64 {
+	if flagEpochTimingRefTimestamp == 0 || flagEpochTimingDuration == 0 {
+		panic("invalid epoch timing config: must specify ALL of --epoch-target-end-time-ref-counter, --epoch-target-end-time-ref-timestamp, and --epoch-target-end-time-duration")
+	}
+	if flagEpochCounter < flagEpochTimingRefCounter {
+		panic("invalid epoch timing config: reference epoch counter must be before root epoch counter")
+	}
```

Comment:
You're right, there is an off-by-one-error in the documentation, it should be clear that we allow equality -- I will update this. 

On the smart contract side, we disallow future reference epochs to avoid needing to account for integer underflow and because I didn't think it would be useful to be able to supply a future reference epoch. The same check is not strictly needed here, but these config values should be equivalent to those used in the smart contract.

Reply:
Updated the error message here: https://github.com/onflow/flow-go/pull/5071/commits/52e78026f0dc8d240dff47f6f76a6841ab110694